shell/platform/linux/fl_key_event_plugin.cc

```diff
@@ -71,6 +77,40 @@ void fl_key_event_plugin_send_key_event(FlKeyEventPlugin* self,
   int64_t scan_code = event->hardware_keycode;
   int64_t unicodeScalarValues = gdk_keyval_to_unicode(event->keyval);
 
+  // Remove lock states from state mask.
+  guint state = event->state & ~(GDK_LOCK_MASK | GDK_MOD2_MASK);
+
+  // GTK keeps track of the state of the locks themselves, not the "down" state
+  // of the key. Flutter expects the "down" state of the modifier key, so we
+  // keep track of the down state here, and send it to the framework. This code
+  // has the flaw that if a key event is missed due to the app losing focus,
+  // then this state will still think the key is down when it isn't, but that is
+  // no worse than for other keys until we implement the sync/cancel events.
+  //
+  // This is necessary to do here instead of in the framework because Flutter
+  // does modifier key syncing in the framework, and will turn on/off these keys
+  // as being "pressed" whenever the lock is on, which breaks a lot of
+  // interactions.
+  //
+  // TODO(gspencergoog): get rid of this tracked state when we are tracking the
+  // state of all keys and sending sync/cancel events when focus is gained/lost.
+  static bool shift_lock_down = false;
```

Comment:
I think GTK offers much more pressing states than these 3, such as alt, scroll lock, etc. Why do we only add these 3?

Reply:
Yes, but only these three are locks. The others are all bits saying if the key is currently pressed, and that's what Flutter is expecting.